tidy3d/components/geometry.py

```diff
@@ -1206,6 +1321,75 @@ def bounds(self) -> Bound:
         coord_max[self.axis] = self.center[self.axis] + self.length / 2.0
         return (tuple(coord_min), tuple(coord_max))
 
+    def volume(self, bounds: Bound) -> float:
+        """Returns object's volume. If the object extends beyound bounds then
+        the computed volume may be only a rough approximation
+
+        Returns
+        -------
+        float
+            Volume.
+        """
+
+        coord_min = self.center[self.axis] - self.length / 2.0
+        coord_max = self.center[self.axis] + self.length / 2.0
+
+        coord_min = max(coord_min, bounds[0][self.axis])
+        coord_max = min(coord_max, bounds[1][self.axis])
+
+        length = coord_max - coord_min
+
+        volume = np.pi * self.radius**2 * length
+
+        # a very loose upper bound on how much of the cylinder is in bounds
```

Comment:
Think there's also a formula for this in the wiki page: https://en.wikipedia.org/wiki/Spherical_cap but if it's not of much importance, dont worry about it.